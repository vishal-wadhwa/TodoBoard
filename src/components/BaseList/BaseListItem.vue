<template>
  <v-container fluid class='pa-1' @click='onItemClickHandler'>
    <v-card v-if='!isNewView'>
      <v-layout
        row
        align-start
        :style='{borderLeftColor: ribbonColor}'
        class='pa-2 bli-ribbon-style'
      >
        <slot name='vf-icon' :type='type'>
          <v-icon>{{type}}</v-icon>
        </slot>
        <v-flex align-self-center class='mt-1 mx-2'>
          <div class='title font-weight-regular'>
            <slot name='vf-title' :title='title' :view='view'>
              <v-card-title class='pa-0'>{{title}}</v-card-title>
            </slot>
          </div>
          <v-divider class='my-2' v-if='!isCompactView'></v-divider>
          <v-card-text text-truncate class='ml-1 pa-0' v-if='!isCompactView'>{{desc}}</v-card-text>
          <v-card-text class='pa-0' v-if='!isCompactView'>
            <slot name='tags' :title='title' :tags='tags'>
              <v-chip
                v-for='(tag, idx) in tags'
                :key='idx'
                small
                label
                :color='tagColor()'
                text-color='white'
              >{{tag}}</v-chip>
            </slot>
          </v-card-text>
        </v-flex>
      </v-layout>
    </v-card>
    <slot v-else name='vn-body' :title='title' :view='view'>
      <v-card flat tile>
        <v-card-actions class='bli-new--border pa-0'>
          <v-btn flat depressed block class='bli-new--height px-0'>
            <slot name='vn-content' :title='title' :view='view'>
              <v-icon>note_add</v-icon>
              {{title}}
            </slot>
          </v-btn>
        </v-card-actions>
      </v-card>
    </slot>
  </v-container>
</template>
<script>
const FULL_VIEW = 'full'
const COMPACT_VIEW = 'compact'
const NEW_VIEW = 'new'
const VIEW_TYPES = [FULL_VIEW, COMPACT_VIEW, NEW_VIEW]
export default {
  data () {
    return {
      colorChoices: ['red', 'green', 'purple', 'indigo', 'primary', 'secondary', 'teal'],
      FULL_VIEW: FULL_VIEW,
      COMPACT_VIEW: COMPACT_VIEW,
      VIEW_TYPES: VIEW_TYPES,
      NEW_VIEW: NEW_VIEW
    }
  },
  props: {
    type: {
      type: String,
      default: 'assignment'
    },
    title: {
      type: String,
      required: true
    },
    desc: {
      type: String
    },
    view: {
      type: String,
      default: NEW_VIEW,
      validator (val) {
        return VIEW_TYPES.includes(val)
      }
    },
    ribbonColor: {
      type: String,
      default: 'black'
    },
    tags: {
      type: Array,
      default: () => []
    }
  },
  methods: {
    tagColor () {
      const idx = Math.random() * this.colorChoices.length >> 0
      return this.colorChoices[idx]
    },
    onItemClickHandler (ev) {
      this.$emit('bli:click', { ...this.$props }, ev)
    }
  },
  computed: {
    isCompactView () {
      return this.view === this.COMPACT_VIEW || ((!this.tags || this.tags.length === 0) && !this.desc)
    },
    isNewView () {
      return this.view === this.NEW_VIEW
    }
  }
}
</script>
<style lang="stylus" scoped>
.bli-new {
  &--border {
    border: 2px dotted gray;
  }

  &--height {
    height: 40px;

    @media screen and (min-width: 960px) {
      height: 80px;
    }
  }
}

.bli-ribbon-style {
  border-left: 3px solid black;
}
</style>
