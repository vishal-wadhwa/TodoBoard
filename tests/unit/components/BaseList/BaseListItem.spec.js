import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import BaseListItem from '@/components/BaseList/BaseListItem'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

describe('BaseListItem.vue', () => {
  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(BaseListItem, opts) : shallowMount(BaseListItem, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  const title = 'Note1'
  const desc = 'random desc'
  const tags = ['Lesson1', 'Lesson2']
  // NEW VIEW TEST
  it('renders title in new view', () => {
    const wrapper = factory({
      propsData: { title },
      localVue
    })
    expect(wrapper.text()).toMatch(title)
  })

  it('does not render desc or tags in new view', () => {
    const wrapper = factory({
      propsData: { title, desc, tags },
      localVue
    })
    expect(wrapper.text()).not.toMatch(desc)
    expect(wrapper.text()).not.toMatch(new RegExp(tags.join('|')))
  })

  // FULL VIEW TEST
  it('renders title in full view', () => {
    const view = 'full'
    const wrapper = factory({
      propsData: { title, view },
      localVue
    })
    expect(wrapper.text()).toMatch(title)
  })

  it('renders desc and tags in full view', () => {
    const view = 'full'
    const wrapper = factory({
      propsData: { view, desc, title, tags },
      localVue
    })
    expect(wrapper.text()).toMatch(desc)
    expect(wrapper.text()).toMatch(new RegExp(tags.join('|')))
  })

  // COMPACT VIEW TEST
  it('renders title in compact view', () => {
    const view = 'compact'
    const wrapper = factory({
      propsData: { view, title },
      localVue
    })
    expect(wrapper.text()).toMatch(title)
  })

  it('does not render desc or tags in compact view', () => {
    const view = 'compact'
    const wrapper = factory({
      propsData: { view, title, tags, desc },
      localVue
    })
    expect(wrapper.text()).not.toMatch(desc)
    expect(wrapper.text()).not.toMatch(new RegExp(tags.join('|')))
  })

  // GENERAL TEST
  it('emits "bli:click" with props data on clicking the container', () => {
    const wrapper = factory({
      propsData: { title, tags, desc },
      localVue
    }, true)

    wrapper.trigger('click')
    const emitObj = wrapper.emitted('bli:click')
    expect(emitObj).toBeTruthy()
    expect(emitObj[0][0]).toEqual(wrapper.props())
  })

  it('emits "bli:delete" when delete button is clicked', () => {
    const view = 'full'
    const wrapper = factory({
      propsData: { title, tags, desc, view },
      localVue
    }, true)

    wrapper.find({ name: 'v-btn' }).trigger('click')

    const emitObj = wrapper.emitted('bli:delete')
    expect(emitObj).toBeTruthy()
    expect(emitObj.length).toEqual(1)
    expect(emitObj[0][0]).toBeInstanceOf(Event)
  })

  it('shows delete button following mouseover/mouseout', async () => {
    const view = 'full'
    const wrapper = factory({
      propsData: { title, tags, desc, view },
      stubs: ['v-card-text', 'v-card-actions']
    }, true)

    const delBtn = wrapper.find({ name: 'v-btn' })

    expect(delBtn.isVisible()).toBeFalsy()
    expect(wrapper.vm.showDeleteIcon).toBeFalsy()

    wrapper.trigger('mouseover')
    await wrapper.vm.$nextTick()

    expect(delBtn.isVisible()).toBeTruthy()
    expect(wrapper.vm.showDeleteIcon).toBeTruthy()

    wrapper.trigger('mouseout')
    await wrapper.vm.$nextTick()

    expect(delBtn.isVisible()).toBeFalsy()
    expect(wrapper.vm.showDeleteIcon).toBeFalsy()
  })

  it('should match full view snapshot', () => {
    const view = 'full'
    const wrapper = factory({
      propsData: { title, tags, desc, view },
      methods: {
        tagColor: () => 'primary'
      }
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match new view snapshot', () => {
    const view = 'new'
    const wrapper = factory({
      propsData: { title, tags, desc, view }
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match compact view snapshot', () => {
    const view = 'compact'
    const wrapper = factory({
      propsData: { title, tags, desc, view }
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
