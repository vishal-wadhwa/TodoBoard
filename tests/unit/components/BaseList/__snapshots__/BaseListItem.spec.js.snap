// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseListItem.vue should match compact view snapshot 1`] = `
<div class="container pa-1 fluid">
  <div class="v-card theme--light">
    <div class="layout pa-2 bli-ribbon-style row align-start" style="border-left-color: black;"><i aria-hidden="true" class="v-icon material-icons theme--light">assignment</i>
      <div class="flex mx-2 align-self-center">
        <div class="flex title font-weight-regular">
          <div class="v-card__title pa-0">
            Note1
            <button type="button" class="ma-0 v-btn v-btn--absolute v-btn--flat v-btn--icon v-btn--right v-btn--small theme--light black--text" style="display: none;">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light" style="font-size: 16px;">close</i></div>
            </button></div>
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseListItem.vue should match full view snapshot 1`] = `
<div class="container pa-1 fluid">
  <div class="v-card theme--light">
    <div class="layout pa-2 bli-ribbon-style row align-start" style="border-left-color: black;"><i aria-hidden="true" class="v-icon material-icons theme--light">assignment</i>
      <div class="flex mx-2 align-self-center">
        <div class="flex title font-weight-regular">
          <div class="v-card__title pa-0">
            Note1
            <button type="button" class="ma-0 v-btn v-btn--absolute v-btn--flat v-btn--icon v-btn--right v-btn--small theme--light black--text" style="display: none;">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light" style="font-size: 16px;">close</i></div>
            </button></div>
        </div>
        <hr class="my-2 v-divider theme--light">
        <div text-truncate="" class="v-card__text ml-1 pa-0">random desc</div>
        <div class="v-card__text pa-0"><span tabindex="0" class="v-chip v-chip--label v-chip--small theme--light primary white--text"><span class="v-chip__content">Lesson1</span></span><span tabindex="0" class="v-chip v-chip--label v-chip--small theme--light primary white--text"><span class="v-chip__content">Lesson2</span></span></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseListItem.vue should match new view snapshot 1`] = `
<div class="container pa-1 fluid">
  <div class="v-card v-card--flat v-card--tile theme--light">
    <div class="v-card__actions bli-new--border pa-0"><button type="button" class="bli-new--height px-0 v-btn v-btn--block v-btn--flat theme--light">
        <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light">note_add</i>
          Note1
        </div>
      </button></div>
  </div>
</div>
`;
